### PR TITLE
Return units in persistence extension commands and support future persisted states

### PR DESCRIPTION
**Pull request by mherwege**
Closes https://github.com/openhab/openhab-core/issues/3732
Closes https://github.com/openhab/openhab-core/issues/3896

This PR implements unit support in all PersistenceExtension commands, returning the full value including unit for Number item types with a dimension. It therefore changes return types from DecimalType to a State type, whereby the state can include the unit.
This has been tested to work with rules DSL, and it should be possible to extend other scripting languages to support this, making calculations with historic states of items with dimensions more straightforward.
This may require adapting user rules and is therefore a breaking change.

Secondly, this PR makes sure persistence extensions work properly with persisted items in the future:
- all `Between` commands have been adapted to be able to properly work across the `now` boundary,
- a series of `Untill` commands have been added, equivalent to the `Since` commands.
- `nexUpdate` and `nexState` commands have been added.
- the `historicState` command has been deprecated (but still available for backward compatibility) and replaced with `persistedState` also accepting future dates.
- the `evolutionRate` command has been renamed to `evolutionRateSince`, `evolutionRateUntill` and `evolutionRateBetween`, bringing it in line with the other commands. This change is required to allow one date signatures to differentiate between `Since` and `Untill`. `evolutionRate` is deprecated and kept for backward compatibility, and interpreted as `evolutionRateSince`.
- added `removeAllStates` `Since`, `Untill` and `Between` commands.

The code now also has full null annotations and checks. In doing so, changes were required to avoid potential null issues. Many of these null issues where due to null being interpreted as now for end dates before, which is not possible when time can be in the future.

The test classes where extended and refactored to also test for all future commands.

Signed-off by: Mark Herwege mark.herwege@telenet.be

### CODE REVIEW

**Comment by J-N-K** [comment]
I understand what you want to achieve, but why did you chose `HistoricItem` as return value? The timestamp has no meaning for averages or similar, a simple `State` would be fine (and could be a `DecimalType` or a `QuantityType` (depending on the item's unit).

**Comment by mherwege** [comment]
> I understand what you want to achieve, but why did you chose `HistoricItem` as return value? The timestamp has no meaning for averages or similar, a simple `State` would be fine (and could be a `DecimalType` or a `QuantityType` (depending on the item's unit).

HistoricItem was already used for minimum, maximum and previous, so wanted to stick to that. But I agree the date returned in the HistoricItem is arbitrary. I will adjust.

**Comment by mherwege** [comment]
@J-N-K the methods now return State.

As this is API breaking anyway, I also aligned method names. EvolutionRate whas the only group of methods that did not have a distinction between Since and Between in the name. I believe this is a rarely used method anyway and I think it is cleaner to have them aligned. If you feel differently, I can revert this last commit.

**Comment by florian-h05** [comment]
Any chance to get this merged for openHAB 4.1 or is that something for a new major release because it's breaking?

**Comment by kaikreuzer** [comment]
@J-N-K As you had commented before, I hoped that you continue on this PR - any chance to do so?

> This may require adapting user rules and is therefore a breaking change.

Breaking user rules is never nice. I'm just trying to get my head around how much this might break. Do you have a typical example where user's will have to adapt their rules? Are these rather exceptional cases or do most rules break by this?
Is there any migration path that we could document as a "quick fix"?
I am not 100% sure whether we should merge this now so shortly before the release. Waiting for a next major release isn't really an option either, that will still take a while...

**Comment by florian-h05** [comment]
If this gets merged for 4.1, we should try to implement a solution for #3896 for 4.1 as well to avoid having multiple sets of breaking changes.

**Comment by J-N-K** [comment]
I have a rule that calculates the mean temperature over the last three days. It uses averageSince and then adds the unit before posting the update to the item.

I still think this is a good addition, but I have no idea how to make it "less breaking".

**Comment by mherwege** [comment]
> I have a rule that calculates the mean temperature over the last three days. It uses averageSince and then adds the unit before posting the update to the item.

Similar for me, and that was the reason to create this PR. It forces you to add the unit again after retrieving the average. That means you need to 'know' the unit in the rule when retrieving to put the right unit back and it goes against the principle of UOM where you don't care.

**Comment by florian-h05** [comment]
> If we all agree we want to do this, I could extend this PR to include a solution for #3896.

I agree, but I'm no core maintainer ;-)
Thank you for the offer! 

> Scripting languages may also be affected. Would any change be required in the javascript helper libraries?

Yes, for renamed methods there would be changes required, however those are minor as the library is only wrapping `PersistenceExtensions` inside a JS class to return JS datatypes.
I would also keep the old ones in the library and log a deprecation warning, so this is less breaking.

For the new functionalities, I have to think about how to integrate them in the library, but I already have an idea.

I can't speak for jRuby, but I guess HABApp is not affected because it is using the REST API to communicate with openHAB and not any Java calls.

**Comment by kaikreuzer** [comment]
> If we all agree we want to do this

I agree that I want this change, I am just not sure when (before or after 4.1). I tend towards waiting after the release, so that we can get some feedback by snapshot and milestone users first and maybe come up with a good migration plan for 4.2.

**Comment by florian-h05** [comment]
Yeah I've already asked myself when this should be implemented.

On one hand, it would be nice to provide the right "tools" together with the new time-series support, 
but on the other hand 4.0 had several breaking changes (and so it would be "nice" to have less/no breaking changes in 4.1) and 1.5 weeks till the 4.1 code-freeze is also a bit short to coordinate the changes across all repos and test this.

**Comment by J-N-K** [comment]
> Similar for me, and that was the reason to create this PR. It forces you to add the unit again after retrieving the average. That means you need to 'know' the unit in the rule when retrieving to put the right unit back and it goes against the principle of UOM where you don't care.

Full ack. But nevertheless all existing rules using persistence extensions need to be changed.

**Comment by spacemanspiff2007** [comment]
> Similar for me, and that was the reason to create this PR. It forces you to add the unit again after retrieving the average.

But why? I thought when the unit is properly set you can just post an empty value and the unit from the metadata will be assumed? Imho you should just be able to post the unitless value and everything should work as expected.

**Comment by mherwege** [comment]
> But why? I thought when the unit is properly set you can just post a unit with an empty value and the unit from the metadata will be assumed? Imho you should just be able to post the unitless value and everything should work as expected.
> The whole point of explicitly providing the unit in the item definition is that there are no ambiguites and the unit can always properly be resolved.

That's true if you immediately post an update to an item. Sometimes, these are intermediate steps in calculations. When you retrieve something from persistence, you have to know what unit it is, and you need to consider that in each step in the calculation, convert the value if necessary yourself as it may be in the wrong unit, and then post an update to an item with the value in the unit the item expects. So the rule needs to know in each step what unit it is working in, but it is not explicit. I prefer to have the Quantities with units in the calculations all the way, to avoid all of this. That way the rule does not have to know the unit it retrieves from persistence and the unit of the item it writes to.

**Comment by mherwege** [comment]
> I agree that I want this change, I am just not sure when (before or after 4.1). I tend towards waiting after the release, so that we can get some feedback by snapshot and milestone users first and maybe come up with a good migration plan for 4.2.

I will work on extending this with logic for future states with target 4.2 for the whole PR.

**Comment by spacemanspiff2007** [comment]
> That's true if you immediately post an update to an item that has the same unit in metadata as the unit for the item you retrieve from persistence.

That's exactly the example @J-N-K made above and that's why I was confused because this should already work.

**Comment by J-N-K** [comment]
My example was just a rule that needs adjustment, not an example to justify this change :-)

**Comment by J-N-K** [comment]
What's the state here? Are the discussed extensions already included and we can review and merge? The sooner, the better: more time for testing.

**Comment by mherwege** [comment]
> What's the state here? Are the discussed extensions already included and we can review and merge? The sooner, the better: more time for testing.

Sorry, not yet. And I won’t be able to work on it for the next few days.

**Comment by mherwege** [comment]
@J-N-K, @florian-h05 I completed the work on persistence extensions, so this is now ready for review.
I messed up my local git history, so had to force push a new state with the full content of the file. I am sorry about that.

**Review by florian-h05 — COMMENTED** [review]
API looks good, thanks!

I will try to test this soon.

**Comment by jimtng** [comment]
Sorry to be so late to the party. I would suggest calling the methods `Until` instead of `Till`.

**Comment by mherwege** [comment]
> Sorry to be so late to the party. I would suggest calling the methods `Until` instead of `Till`.

Fine for me, change done.

**Comment by jimtng** [comment]
> Fine for me, change done.

Thanks so much! The OCD in me would love the "till" in the comments also updated to say "until" to be more consistent.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:2376]: This would be API breaking. HistoricItem a a State with a ZonedDateTime. It is returned when the ZonedDateTime makes sense. My original proposal was to use HistoricItem everywhere, but @J-N-K correctly reacted that it does not make sense for many of the methods. So I kept HistoricItem where it was already returned (and it makes sense) and use State where the ZonedDateTime does not make sense.

Comment (jimtng) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:2371]: No, I didn't mean to return HistoricItem everywhere. 

I mean that for methods that return a HistoricItem (such as maximumSince), or an Iterable<HistoricItem> (such as getAllStatesSince), make sure that the historicItem.getState() would return a QuantityType instead of DecimalType where applicable.

So from an API consumer's perspective:

```java
HistoricItem historyMax = PersistenceExtensions.maximumSince(myItem, ZonedDateTime.now().minusDays(1));
var max = historyMax.getState(); // <= this will be a QuantityType instead of DecimalType
```

Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:2371]: Unless I am missing something, I believe that is case already. There are also a series of tests testing these methods with QuantityTypes.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java]: Done

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (jimtng) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:2371]: My apologies, I didn't look at the tests, you're correct. Where did the QuantityType come from? Did it come from the persistence service?

Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:2371]: Yes, indeed.

**Comment by mherwege** [comment]
I am going through the javadocs again. Many of them need updates or clarifications. I am working on that and will come back with refinements on them.

**Comment by mherwege** [comment]
> I am going through the javadocs again. Many of them need updates or clarifications. I am working on that and will come back with refinements on them.

Done, and removed some commented code and solved an issue.

**Comment by jimtng** [comment]
> Scripting languages may also be affected. Would any change be required in the javascript helper libraries? What about jruby

If the old `evolutionRate()` were to remain, perhaps with a deprecation warning like `historicState`, then it wouldn't break any jruby scripts, and perhaps also other scripting languages?

**Comment by mherwege** [comment]
> If the old `evolutionRate()` were to remain, perhaps with a deprecation warning like `historicState`, then it wouldn't break any jruby scripts, and perhaps also other scripting languages?

Done that now. A warning similar to `historicState` will be logged.

**Comment by jimtng** [comment]
If this is not going to be API Breaking, the label should be removed.

**Comment by J-N-K** [comment]
I this now ready for review?

**Comment by mherwege** [comment]
> I this now ready for review?

It is from my perspective. I already addressed a number of review comments from other people.

**Comment by florian-h05** [comment]
@openhab/core-maintainers Can you please review this PR? 
It should not be merged too late before the 4.2 release as we also need some time to adjust the helper libraries and the documentation, so the earlier the better.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (J-N-K) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:99]: Shouldn't this result in a warning similar to the one on l. 111? ("No default persistence service configured")

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (J-N-K) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:142]: see above

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (J-N-K) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:280]: Why is that needed? If `service` is an instance of `QueryablePersistenceService` is is not `null` (also applies to other methods).

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (J-N-K) [inline bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:292]: I would prefer to keep the `else` branch, it is more clear (also applies to other methods).

**Review by J-N-K — CHANGES_REQUESTED** [review]
In general looks very good to me. This is a tremendous work and makes the persistence extensions more consistent and usable.

I have left some minor comments and can you resolve the merge conflict? Thanks.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:99]: `getDefaultServiceId()` already logs a warning.

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:292]: OK

File: bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java
Comment (mherwege) [reply bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java:280]: Indeed, not required.

**Comment by mherwege** [comment]
@J-N-K I rebased and made some changes relative to the review comments.

**Review by J-N-K — APPROVED** [review]
Thanks! Can you als add a PR for the breaking-change notice in openhab-distro and adjust the documentation on okpenhab-docs?

**Comment by florian-h05** [comment]
FYI openHAB JavaScript library PR to adjust to this PR is already created: https://github.com/openhab/openhab-js/pull/331
Blockly needs to be updated next ...

**Comment by lolodomo** [comment]
@mherwege : does it require a note in the release notes to mention a breaking change ?

**Comment by J-N-K** [comment]
Already done:https://github.com/openhab/openhab-distro/pull/1652

**Comment by jimtng** [comment]
core build seems to be on a daily basis

**Comment by jimtng** [comment]
openhab-core build is done now. Next, we need the distro build to kick in

**Comment by J-N-K** [comment]
Distro-build should not be required to build webui

**Comment by jimtng** [comment]
I've just noticed that the "between" variant of the old `evolutionRate()` was removed in this PR, instead of deprecated. Should we add it back for the sake of compatibility?

**Comment by mherwege** [comment]
> I've just noticed that the "between" variant of the old `evolutionRate()` was removed in this PR, instead of deprecated. Should we add it back for the sake of compatibility?

Yes, makes sense. I missed that one.

**Comment by jimtng** [comment]
Another change I've just noticed: Previously I could call any methods with a null serviceid. 

Now, due to `@NonNullByDefault` on the `PersistenceExtensions` class, and the absence of `@Nullable` in front of serviceId, this is no longer possible.

This change has caused some errors in existing scripts. Should we restore the previous behaviour?

**Comment by jimtng** [comment]
@mherwege I edited/updated the [post above](https://github.com/openhab/openhab-core/pull/3736#issuecomment-2093140530), in case you've only read the original message.

**Comment by mherwege** [comment]
> Another change I've just noticed: Previously I could call any methods with a null serviceid to indicate that the default persistence service is to be used.

It has never been documented it could be used that way. Also, quickly looking through the old code again, it doesn't look like it worked for all actions, and it even gave wrong result for some (part of the query working and another part returning null and defaulting to a wrong default value, so the combination being wrong).
While it is certainly doable to introduce this, I am wondering if it is worth doing this to fix wrong usage of the actions in the first place.

**Comment by mherwege** [comment]
@jimtng PR is created to allow null serviceId.

**Comment by jimtng** [comment]
> @jimtng PR is created to allow null serviceId.


Thanks! I was pondering about your previous response. I agree that it wasn't documented in the javadocs. I've started working on modifications in the jruby helper library to avoid using a null serviceId. However, I think it's a handy thing to support but either way is fine.

**Comment by njordan77** [comment]
Hello, i changed to M3 and now a rule does throw an error based on this line. Says "unknown variable or line" pointint to the "minus"... Any idea what i have to use instead? Thanks Norbert
`
var AvailChargePower = (miFro_ACpower.averageSince(now.minusMinutes(2)) - 400)`

**Pull request merged**
